File: leetcode/201-300/284-peeking-iterator/bench/peekiter.c
/* LeetCode 284 bench mirror — C. Same eager wrapper, same two-peeks-per-next
 * mix, same sink plus pull count. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#define N 200000
#define ROUNDS 320
typedef struct { int64_t *data; int64_t len, pos, pulls; } Source;
static int src_has_next(Source *s) { return s->pos < s->len; }
static int64_t src_next(Source *s) { int64_t v = s->data[s->pos]; s->pos++; s->pulls++; return v; }
typedef struct { Source src; int64_t slot; int full; } Peeking;
static Peeking make_peeking(int64_t *data, int64_t n) {
    Peeking p;
    p.src.data = malloc((size_t)n*sizeof(int64_t));
    memcpy(p.src.data, data, (size_t)n*sizeof(int64_t));
    p.src.len = n; p.src.pos = 0; p.src.pulls = 0;
    p.slot = 0; p.full = 0;
    if (src_has_next(&p.src)) { p.slot = src_next(&p.src); p.full = 1; }
    return p;
}
static int64_t peek(Peeking *p) { return p->slot; }
static int has_next(Peeking *p) { return p->full; }
static int64_t next_(Peeking *p) {
    int64_t v = p->slot;
    if (src_has_next(&p->src)) p->slot = src_next(&p->src); else p->full = 0;
    return v;
}
int main(void) {
    int64_t *data = malloc((size_t)N*sizeof(int64_t));
    int64_t seed = 20260823;
    for (int64_t i = 0; i < N; i++) { seed = (seed*1103515245LL+12345LL)%2147483648LL; data[i] = seed % 100003LL; }
    int64_t sink = 0, total = 0;
    for (int r = 0; r < ROUNDS; r++) {
        Peeking p = make_peeking(data, N);
        int64_t h = 0, pos = 1;
        while (has_next(&p)) {
            h = (h*31 + peek(&p)*pos) % 1000000007LL;
            h = (h*31 + peek(&p)) % 1000000007LL;
            int64_t v = next_(&p);
            h = (h*31 + v) % 1000000007LL;
            pos++;
        }
        total += p.src.pulls;
        free(p.src.data);
        sink = (sink + h) % 1000000007LL;
    }
    printf("%lld %lld\n", (long long)sink, (long long)total);
    return 0;
}

File: leetcode/201-300/284-peeking-iterator/bench/peekiter.rs
// LeetCode 284 bench mirror — Rust. Same eager wrapper and operation mix.
const N: usize = 200000;
const ROUNDS: usize = 320;
struct Source { data: Vec<i64>, pos: usize, pulls: i64 }
impl Source {
    fn has_next(&self) -> bool { self.pos < self.data.len() }
    fn next(&mut self) -> i64 { let v = self.data[self.pos]; self.pos += 1; self.pulls += 1; v }
}
struct Peeking { src: Source, slot: i64, full: bool }
fn make_peeking(data: &[i64]) -> Peeking {
    let mut src = Source { data: data.to_vec(), pos: 0, pulls: 0 };
    let (mut slot, mut full) = (0i64, false);
    if src.has_next() { slot = src.next(); full = true; }
    Peeking { src, slot, full }
}
impl Peeking {
    fn peek(&self) -> i64 { self.slot }
    fn has_next(&self) -> bool { self.full }
    fn next(&mut self) -> i64 {
        let v = self.slot;
        if self.src.has_next() { self.slot = self.src.next(); } else { self.full = false; }
        v
    }
}
fn main() {
    let mut seed: i64 = 20260823;
    let mut data = vec![0i64; N];
    for i in 0..N { seed = (seed*1103515245+12345)%2147483648; data[i] = seed % 100003; }
    let (mut sink, mut total) = (0i64, 0i64);
    for _ in 0..ROUNDS {
        let mut p = make_peeking(&data);
        let (mut h, mut pos) = (0i64, 1i64);
        while p.has_next() {
            h = (h*31 + p.peek()*pos) % 1000000007;
            h = (h*31 + p.peek()) % 1000000007;
            let v = p.next();
            h = (h*31 + v) % 1000000007;
            pos += 1;
        }
        total += p.src.pulls;
        sink = (sink + h) % 1000000007;
    }
    println!("{} {}", sink, total);
}
